### DISCUSSION TITLE
How to smartly invoke methods of different types?

### DISCUSSION THREAD

**Discussion by aksgupta98** [uber-go/fx#1041] [Q&A]
Hi everyone, I am building a go application which has multiple modules. Each module has a different service hence I do not want to clutter my code by invoking every service before using it.
Please suggest a better way to achieve this

My Current Code looks like this - 

```
package factory

var riderServ *riderService.RiderService
var userServ *userService.UserService

func GetFactory() {
	fx.New(
		userService.Module,
		userDaoImpl.Module,
		riderService.Module,
		fx.Invoke(getRiderServiceObj),
		fx.Invoke(getUserServiceObj),
	)
	riderServ.DummyRiderService()
	userServ.DummyUserService()
}

func getRiderServiceObj(a *riderService.RiderService) {
	riderServ = a
}

func getUserServiceObj(b *userService.UserService) {
	userServ = b
}
```

I have used Invoke multiple times to get a instance of a service, and I am sure there is a better way to do it.

**Comment by sywhang** [comment]
You may want to check out our docs.

First, Invoke shouldn't be used for doing things like this - see: https://uber-go.github.io/fx/modules.html#invoke-sparingly. 

For the purpose of initializing a variable that was provided to the DI container, you can use [fx.Populate]https://pkg.go.dev/go.uber.org/fx#Populate) to extract the value. That being said, the whole point of Fx is to avoid having to initialize your "global variables" this way. You may benefit from using Lifecycles for managing the lifecycles of your components: https://uber-go.github.io/fx/lifecycle.html
